rslib/src/image_occlusion/imageocclusion.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+// split following
+// text = "rect:399.01,99.52,167.09,33.78:fill=#0a2cee:stroke=1"
+// with
+// result = "data-shape="rect" data-left="399.01" data-top="99.52" data-width="167.09" data-height="33.78" data-fill="\#0a2cee" data-stroke="1""
+pub fn get_image_cloze_data(text: &str) -> String {
+    let mut result = String::new();
+    let mut shape = "";
+    let mut angle = "";
+    let mut fill = "";
+    let mut height = "";
+    let mut hideinactive = "";
+    let mut left = "";
+    let mut points = "";
+    let mut qmask = "";
+    let mut rx = "";
+    let mut ry = "";
+    let mut top = "";
+    let mut width = "";
+
+    let parts: Vec<&str> = text.split(':').collect();
+
+    if parts.len() >= 2 {
+        shape = parts[0];
+        for part in parts[1..].iter() {
+            let values: Vec<&str> = part.split('=').collect();
+            if values.len() >= 2 {
+                match values[0] {
+                    "left" => left = values[1],
+                    "top" => top = values[1],
+                    "width" => width = values[1],
+                    "height" => height = values[1],
+                    "fill" => fill = values[1],
+                    "rx" => rx = values[1],
+                    "ry" => ry = values[1],
+                    "angle" => angle = values[1],
+                    "points" => points = values[1],
+                    "qmask" => qmask = values[1],
+                    "hideinactive" => hideinactive = values[1],
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    if !shape.is_empty() {
+        result.push_str(&format!("data-shape=\"{}\" ", shape));
+    }
+
+    // special properties to shapes
+    match shape {
+        "ellipse" => {
+            if !rx.is_empty() {
+                result.push_str(&format!("data-rx=\"{}\" ", rx));
+            }
+            if !ry.is_empty() {
+                result.push_str(&format!("data-ry=\"{}\" ", ry));
+            }
+        }
+
+        "polygon" => {
+            if !points.is_empty() {
+                let points_vec: Vec<&str> = points.split(' ').collect();
+                let mut point_str = String::new();
+                for point in &points_vec {
+                    let xy: Vec<&str> = point.split(',').collect();
+                    point_str.push_str(&format!("[{},{}],", xy[0], xy[1]));
```

Comment:
This will crash if a point pair is missing a comma. Here's a more efficient version that will skip invalid entries:

```diff
diff --git a/rslib/src/image_occlusion/imageocclusion.rs b/rslib/src/image_occlusion/imageocclusion.rs
index 0f4a12902..f2c3b41ee 100644
--- a/rslib/src/image_occlusion/imageocclusion.rs
+++ b/rslib/src/image_occlusion/imageocclusion.rs
@@ -1,6 +1,8 @@
 // Copyright: Ankitects Pty Ltd and contributors
 // License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

+use std::fmt::Write;
+
 // split following
 // text = "rect:399.01,99.52,167.09,33.78:fill=#0a2cee:stroke=1"
 // with
@@ -52,12 +54,12 @@ pub fn get_image_cloze_data(text: &str) -> String {
                     }
                     "points" => {
                         if !values[1].is_empty() {
-                            let points_vec: Vec<&str> = values[1].split(' ').collect();
                             let mut point_str = String::new();
-                            for point in &points_vec {
-                                let xy: Vec<&str> = point.split(',').collect();
-                                point_str.push_str(&format!("[{},{}],", xy[0], xy[1]));
+                            for point_pair in values[1].split(' ') {
+                                let Some((x, y)) = point_pair.split_once(',') else { continue };
+                                write!(&mut point_str, "[{},{}],", x, y).unwrap();
                             }
+                            // remove the trailing comma
                             point_str.pop();
                             if !point_str.is_empty() {
                                 result.push_str(&format!("data-points=\"[{}]\" ", point_str));
```

Reply:
Thanks